File: ruby/anagram/anagram.rb
class Anagram
  def initialize(subject)
    @subject = AnagramWord.new(subject)
  end

  def match(words)
    words.select do |word|
      subject == AnagramWord.new(word)
    end
  end

  private

  attr_reader :subject
end

class AnagramWord
  def initialize(word)
    @word = word
  end

  def ==(other)
    !same?(other) && same_chars?(other)
  end

  protected

  def normalize
    word.downcase
  end

  def sorted_chars
    normalize.chars.sort
  end

  private

  attr_reader :word

  def same?(other)
    normalize == other.normalize
  end

  def same_chars?(other)
    sorted_chars == other.sorted_chars
  end
end

File: elixir/anagram/anagram.exs
defmodule Anagram do
  def match(subject, words) do
    normalized_subject = normalize(subject)

    Enum.filter(words, anagram?(normalized_subject, &1))
  end

  defp anagram?(normalized_subject, word) do
    normalized_subject == normalize(word)
  end

  defp normalize(word) do
    String.codepoints(word)
    |> Enum.sort
  end
end
